File: LeonEdBackend/Controllers/FeeController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using LeonEdBackend.DTOs.FeeDtos;
using LeonEdBackend.Services.Interfaces;

namespace LeonEdBackend.Controllers
{
    [Route("api/[controller]")]
    [Authorize(Roles = "SchoolAdmin")]
    public class FeeController : BaseController
    {
        private readonly IFeeService _feeService;

        public FeeController(IFeeService feeService)
        {
            _feeService = feeService;
        }

        /// <summary>Record a fee payment for a student</summary>
        [HttpPost("record")]
        public async Task<IActionResult> RecordPayment([FromBody] RecordFeePaymentRequest request)
        {
            var schoolId = RequireSchoolId();
            var result = await _feeService.RecordPayment(schoolId, request);
            return result.Success ? Ok(result) : BadRequest(result);
        }

        /// <summary>Manually clear a student's fees for a term</summary>
        [HttpPost("clear/{studentId}/{termId}")]
        public async Task<IActionResult> ClearStudent(Guid studentId, Guid termId)
        {
            var schoolId = RequireSchoolId();
            var userId = GetUserId();
            var result = await _feeService.ClearStudent(schoolId, studentId, termId, userId);
            return result.Success ? Ok(result) : BadRequest(result);
        }

        /// <summary>Check fee status for a student in a term</summary>
        [HttpGet("student/{studentId}/term/{termId}")]
        public async Task<IActionResult> GetStudentFeeStatus(Guid studentId, Guid termId)
        {
            var schoolId = RequireSchoolId();
            var result = await _feeService.GetStudentFeeStatus(schoolId, studentId, termId);
            return result.Success ? Ok(result) : BadRequest(result);
        }

        /// <summary>Get fee overview for all students in a class for a term</summary>
        [HttpGet("class/{classId}/term/{termId}")]
        public async Task<IActionResult> GetClassFeeOverview(Guid classId, Guid termId)
        {
            var schoolId = RequireSchoolId();
            var result = await _feeService.GetClassFeeOverview(schoolId, classId, termId);
            return result.Success ? Ok(result) : BadRequest(result);
        }
    }
}

File: src/controllers/fee.controller.ts
import { Response, NextFunction } from "express";
import { AuthenticatedRequest } from "../types";
import { FeeService } from "../services/fee.service";
import { failResponse } from "../utils/response";

export class FeeController {
  static async recordPayment(req: AuthenticatedRequest, res: Response, next: NextFunction) {
    try {
      const schoolId = req.schoolId!;
      const result = await FeeService.recordPayment(schoolId, req.body);
      if (result.success) {
        return res.status(200).json(result);
      }
      return res.status(400).json(result);
    } catch (error) {
      next(error);
    }
  }

  static async clearStudent(req: AuthenticatedRequest, res: Response, next: NextFunction) {
    try {
      const schoolId = req.schoolId!;
      const userId = req.user?.id!;
      const { studentId, termId } = req.params;

      const result = await FeeService.clearStudent(schoolId, studentId, termId, userId);
      if (result.success) {
        return res.status(200).json(result);
      }
      return res.status(400).json(result);
    } catch (error) {
      next(error);
    }
  }

  static async getStudentFeeStatus(req: AuthenticatedRequest, res: Response, next: NextFunction) {
    try {
      const schoolId = req.schoolId!;
      const { studentId, termId } = req.params;

      const result = await FeeService.getStudentFeeStatus(schoolId, studentId, termId);
      if (result.success) {
        return res.status(200).json(result);
      }
      return res.status(400).json(result);
    } catch (error) {
      next(error);
    }
  }

  static async getClassFeeOverview(req: AuthenticatedRequest, res: Response, next: NextFunction) {
    try {
      const schoolId = req.schoolId!;
      const { classId, termId } = req.params;

      const result = await FeeService.getClassFeeOverview(schoolId, classId, termId);
      if (result.success) {
        return res.status(200).json(result);
      }
      return res.status(400).json(result);
    } catch (error) {
      next(error);
    }
  }

  static async uploadReceipt(req: AuthenticatedRequest, res: Response, next: NextFunction) {
    try {
      const schoolId = req.schoolId!;
      const userId = req.user?.id!;
      const result = await FeeService.uploadReceipt(schoolId, userId, req.body);
      if (result.success) {
        return res.status(200).json(result);
      }
      return res.status(400).json(result);
    } catch (error) {
      next(error);
    }
  }

  static async getMyFeeStatus(req: AuthenticatedRequest, res: Response, next: NextFunction) {
    try {
      const schoolId = req.schoolId!;
      const userId = req.user?.id!;
      const termId = req.query.termId as string;

      if (!termId) {
        return res.status(400).json(failResponse("Missing termId in query parameter."));
      }

      const result = await FeeService.getMyFeeStatus(schoolId, userId, termId);
      if (result.success) {
        return res.status(200).json(result);
      }
      return res.status(400).json(result);
    } catch (error) {
      next(error);
    }
  }
}
